### ISSUE TITLE
Vulnerable to Man-In-The-Middle (MITM) attacks

### ISSUE THREAD

**Issue by stevenjohnstone** [luarocks/luarocks#1215]
I installed Luarocks 3.3.1 on Debian as per instructions [here](https://github.com/luarocks/luarocks/wiki/Installation-instructions-for-Unix).

I modified my /etc/hosts so that luarocks.org would resolve to a local server. The local server has a self-signed certificate:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            3e:e8:91:9b:81:94:cb:9a:81:06:ae:ed:01:b5:3f:45:4f:19:59:c8
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
        Validity
            Not Before: Aug  2 11:48:59 2020 GMT
            Not After : Aug  2 11:48:59 2021 GMT
        Subject: C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
... etc etc
```

Downloads and searches of https://luarocks.org didn't result in any warnings or errors from luarocks. This indicates that

1. the certificate was not validated against root CAs for the system
2. the hostname "luarocks.org" was not checked to be in the certificate SANs or CN

Looking at the alternatives to luasec transport, curl and wget, it appears that those have certificate validation disabled as well.

In summary, in the default configuration (and likely with any other configuration options), Luarocks is vulnerable to MITM attacks.

Thanks to @daurnimator for answering questions on this.

**Comment by stevenjohnstone** [comment]
I can have a stab at fixing this.

Broad strokes...

Luasec can be made to do the right thing on Debian without much fuss e.g. something along these lines (with better taste)

```lua
local socket = require("socket")
local ssl = require("ssl")

local params = {
   mode = "client",
   protocol = "tlsv1_2",
   capath = "/etc/ssl/certs/",
   verify = {"peer", "fail_if_no_peer_cert"}
}

local luarocks = "luarocks.org"
local san_oid = "2.5.29.17"

local peer = socket.tcp()
peer:connect("luarocks.org", 443)

peer = assert(ssl.wrap(peer, params))
assert(peer:dohandshake())

local found = false
for k, ext in pairs(peer:getpeercertificate():extensions()) do
   if k == san_oid then
      for _, name in ipairs(ext.dNSName) do
         if name == luarocks then
            found = true
         end
      end
   end
end
for _, cn in ipairs(peer:getpeercertificate():subject()) do
   if cn.value == luarocks then
      found = true
   end
end
peer:close()

print(found and "found" or "didn't find", luarocks, " in certificate")
```

Other distros put their certs elsewhere but that can be detected at runtime. A handy list of locations for the most popular distros can be found here: https://github.com/golang/go/blob/master/src/crypto/x509/root_linux.go.

OSX and windows would need some surgery on Luasec to import the root CAs from the keychains. There's plenty of prior art on how to do this in, for example, https://github.com/golang/go/tree/master/src/crypto/x509.

Some other light reading: https://wiki.openssl.org/index.php/Hostname_validation

**Comment by daurnimator** [comment]
> ```lua
>    if cn.value == luarocks then
> ```

Wildcard certs are also a thing.
However there are many more validation rules than just host checking.

There is logic for in OpenSSL (for at least all supported versions), I consider it a deficiency in luasec to not expose it.

**Comment by stevenjohnstone** [comment]
> Wildcard certs are also a thing.

Absolutely: making no claims at completeness, this is just a quick exploration of what can be done already with luasec without changing it.

> However there are many more validation rules than just host checking.

I think validation against a CA cert and intermediates is covered by the "peer", "fail_if_no_peer_cert" and "capath" options? If not, luasec violates the principle of least surprise in a very upsetting way :) 

The thing I'm driving at is that I *think* it's possible with some lua-side work to get both the certificate chain validation and the hostname checks.

> 
> There is logic for in OpenSSL (for at least all supported versions), I consider it a deficiency in luasec to not expose it.

Yep, in a perfect world, it'd expose the interface X509_check_host(). In an even more perfect world, there'd be a library which would hide all the details of openssl which bleed into Luasec via the "params" table and make it essentially impossible to get this wrong. I think I could make a wrapper around some C which would

1. only offer client functionality (to keep things simple for the user)
2. pick a small set of good ciphers
3. only offer tls 1.2 or later
4. require a hostname on connection so validation can be completed internally
5. load CA certs from wherever makes sense for the OS

Interface like

```lua
local conn, err = tls.connect(hostname)
```
and nothing else. If I were to go to those lengths, I'd probably use BoringSSL

**Comment by daurnimator** [comment]
> Yep, in a perfect world, it'd expose the interface X509_check_host(). I

That function hasn't been recommended in several years.
Since OpenSSL 1.0.2 (released 2015; and end-of-lifed in 2019) `X509_VERIFY_PARAM_set1_host` is a nice easy way to do things.

**Comment by stevenjohnstone** [comment]
> > Yep, in a perfect world, it'd expose the interface X509_check_host(). I
> 
> That function hasn't been recommended in several years.
> Since OpenSSL 1.0.2 (released 2015; and end-of-lifed in 2019) `X509_VERIFY_PARAM_set1_host` is a nice easy way to do things.

Yep. Something, something DANE something, something dark side. I wouldn't recommend any of the OpenSSL API, to be fair :)

**Comment by hishamhm** [comment]
> Looking at the alternatives to luasec transport, curl and wget, it appears that those have certificate validation disabled as well.

I believe that for wget and curl transports you can set `check_certificates = true` in your LuaRocks config file.

**Comment by daurnimator** [comment]
> I believe that for wget and curl transports you can set `check_certificates = true` in your LuaRocks config file.

Why isn't this on by default?

**Comment by hishamhm** [comment]
> Why isn't this on by default?

Good question, I don't recall why it was off by default, when the flag was added I think it was kept `false` by default for backwards compatibility.

**Comment by daurnimator** [comment]
So I think there are a few things that should happen:
  - luasec needs to gain support for checking hostnames
  - luarocks should gain a config var or flag to select the downloader (at the moment if luasec is installed there is no way to *not* use it)
  - luarocks should have `check_certificates = true` by default

**Comment by stevenjohnstone** [comment]
Maybe worth adding a [security advisory](https://docs.github.com/en/free-pro-team@latest/github/managing-security-vulnerabilities/about-github-security-advisories) for this issue?

**Comment by koct9i** [comment]
I guess this was workaround for broken certificate #286
11 years ago.

And certificate right now is still incorrect: issued for "moonscript.org" without wildcard.
https://rocks.moonscript.org

**Comment by mbartlett21** [comment]
For Linux, luasec's https.lua can be overridden to use the system certificate bundle:

```diff
--- https.lua
+++ https.new.lua
@@ -29,6 +29,7 @@
 local cfg = {
   protocol = "any",
   options  = {"all", "no_sslv2", "no_sslv3", "no_tlsv1"},
-  verify   = "none",
+  verify   = "peer",
+  cafile   = "/etc/ssl/certs/ca-bundle.crt", -- or wherever
 }
```

For a Windows install, the bundle can be downloaded from https://curl.se/ca/cacert.pem , or an installed program's one can be reused (for example Git for Windows creates `C:\Program Files\Git\usr\ssl\certs\ca-bundle.crt`)

And once that is done it appears to be a bit better. It still allows connecting to `wrong.host.badssl.com` though (the certificate there is for `*.badssl.com`)
